contracts/crosschain/axelar/AxelarGatewaySource.sol

```diff
@@ -30,29 +41,58 @@ abstract contract AxelarGatewaySource is IERC7786GatewaySource, AxelarGatewayBas
         bytes[] calldata attributes
     ) external payable returns (bytes32 sendId) {
-        require(msg.value == 0, UnsupportedNativeTransfer());
-        // Use of `if () revert` syntax to avoid accessing attributes[0] if it's empty
-        if (attributes.length > 0)
-            revert UnsupportedAttribute(attributes[0].length < 0x04 ? bytes4(0) : bytes4(attributes[0][0:4]));
+        // Process attributes (relay)
+        bool withRelay = false;
+        uint256 value = 0;
+        address refundRecipient = address(0);
+
+        for (uint256 i = 0; i < attributes.length; ++i) {
+            (withRelay, value, , refundRecipient) = ERC7786Attributes.tryDecodeRequestRelayCalldata(attributes[i]);
+            require(withRelay, UnsupportedAttribute(attributes[i].length < 0x04 ? bytes4(0) : bytes4(attributes[i])));
+        }
+        if (!withRelay) {
+            sendId = bytes32(++_sendId);
+        }
+        require(msg.value == value, UnsupportedNativeTransfer());
 
         // Create the package
         bytes memory sender = InteroperableAddress.formatEvmV1(block.chainid, msg.sender);
         bytes memory adapterPayload = abi.encode(sender, recipient, payload);
 
-        // Emit event
-        sendId = bytes32(0); // Explicitly set to 0
+        // Emit event early (stack too deep)
         emit MessageSent(sendId, sender, recipient, payload, 0, attributes);
 
         // Send the message
         (bytes2 chainType, bytes calldata chainReference, ) = recipient.parseV1Calldata();
-        string memory axelarDestination = getAxelarChain(InteroperableAddress.formatV1(chainType, chainReference, ""));
         bytes memory remoteGateway = getRemoteGateway(chainType, chainReference);
-        _axelarGateway.callContract(
-            axelarDestination,
-            address(bytes20(remoteGateway)).toChecksumHexString(), // TODO non-evm chains?
-            adapterPayload
-        );
+        string memory axelarDestination = getAxelarChain(InteroperableAddress.formatV1(chainType, chainReference, ""));
+        string memory axelarTarget = address(bytes20(remoteGateway)).toChecksumHexString(); // TODO non-evm chains?
+
+        _axelarGateway.callContract(axelarDestination, axelarTarget, adapterPayload);
+
+        if (withRelay) {
+            _axelarGasService.payNativeGasForContractCall{value: msg.value}(
+                address(this),
+                axelarDestination,
+                axelarTarget,
+                adapterPayload,
+                refundRecipient
+            );
+        } else {
+            _details[sendId] = MessageDetails(axelarDestination, axelarTarget, adapterPayload);
+        }
+    }
+
+    function requestRelay(bytes32 sendId, uint256 /*gasLimit*/, address refundRecipient) external payable {
+        MessageDetails storage details = _details[sendId];
+        require(details.payload.length > 0);
 
-        return sendId;
+        _axelarGasService.payNativeGasForContractCall{value: msg.value}(
```

Comment:
Is it okay to call this function _after_ `callContract` has been invoked? I have a feeling that it needs to be called beforehand...

Reply:
Source:

https://github.com/axelarnetwork/axelar-cgp-solidity/blob/d030522e97463061a0c5f4c17bfa50c11cf23b59/contracts/gas-service/AxelarGasService.sol#L156

Reply:
Good observation, I'll check that documentation

Reply:
Note: that means the implementation of `sendMessage` -> `requestRelay` should not work. @cjcobb23 what is the interface to pay gas AFTER a message is sent ?

Reply:
As far as I know it's the `addGas` functions and those take tx hash and log index!

It really seems like there's no way to implement `requestRelay` for this adapter other than delaying `callContract` until `requestRelay` is called...